File: ios/HybridImageToPdf.swift
// Bridges HybridImageToPdfSpec to ImageLoader/PdfBuilder; handles output writing.

import Foundation
import NitroModules
import UIKit

// NitroModules bridges errors to JS via `description`, not `errorDescription`.
enum PdfError: Error {
  case invalidImage(String)
  case downloadFailed(String)
  case downloadTimeout(String)
  case writeFailed(String)
  case permissionDenied(String)
  case encryptionFailed(String)
}

extension PdfError: CustomStringConvertible {
  var description: String {
    switch self {
    case .invalidImage(let detail):
      return "[INVALID_IMAGE] \(detail)"
    case .downloadFailed(let detail):
      return "[DOWNLOAD_FAILED] \(detail)"
    case .downloadTimeout(let detail):
      return "[DOWNLOAD_TIMEOUT] \(detail)"
    case .writeFailed(let detail):
      return "[WRITE_FAILED] \(detail)"
    case .permissionDenied(let detail):
      return "[PERMISSION_DENIED] \(detail)"
    case .encryptionFailed(let detail):
      return "[ENCRYPTION_FAILED] \(detail)"
    }
  }
}

extension PdfError: LocalizedError {
  var errorDescription: String? { description }
}

class HybridImageToPdf: HybridImageToPdfSpec {
  func createPdf(options: CreatePdfOptions) throws -> Promise<CreatePdfResult> {
    return Promise.async {
      guard !options.images.isEmpty else {
        throw PdfError.invalidImage("images must be a non-empty array")
      }

      // Load images, downsampling via maxWidth/maxHeight if set.
      let uiImages: [UIImage] = try options.images.map { input in
        do {
          return try ImageLoader.load(
            type: input.type,
            value: input.value,
            maxWidth: options.processing?.maxWidth,
            maxHeight: options.processing?.maxHeight
          )
        } catch let error as ImageLoadError {
          throw Self.mapLoadError(error)
        }
      }

      // Build the PDF (page sizing, fit, margin, metadata, password).
      let built: (data: Data, pageSizes: [CGSize])
      do {
        built = try PdfBuilder.build(
          images: uiImages,
          page: options.page,
          processing: options.processing,
          metadata: options.metadata,
          userPassword: options.security?.password,
          ownerPassword: options.security?.ownerPassword
        )
      } catch let error as PdfBuildError {
        throw PdfError.invalidImage(error.localizedDescription)
      }

      // Write output per output.format/directory/absolutePath.
      let outputFormat = options.output?.format ?? .file
      var writeFile = false
      var writeBase64 = false
      switch outputFormat {
      case .file:
        writeFile = true
      case .base64:
        writeBase64 = true
      case .both:
        writeFile = true
        writeBase64 = true
      }

      var filePath: String?
      var fileSize: Double?

      if writeFile {
        let path = try Self.resolveOutputPath(output: options.output)
        do {
          try built.data.write(to: URL(fileURLWithPath: path), options: .atomic)
        } catch {
          throw PdfError.writeFailed("Failed to write PDF to \(path): \(error.localizedDescription)")
        }
        filePath = path
        fileSize = Double(built.data.count)
      }

      let base64: String? = writeBase64 ? built.data.base64EncodedString() : nil

      let pages = built.pageSizes.map { PdfPageInfo(width: Double($0.width), height: Double($0.height)) }
      return CreatePdfResult(
        filePath: filePath,
        base64: base64,
        numberOfPages: Double(uiImages.count),
        fileSize: fileSize,
        pages: pages
      )
    }
  }

  private static func mapLoadError(_ error: ImageLoadError) -> PdfError {
    switch error {
    case .invalidImage(let detail):
      return .invalidImage(detail)
    case .downloadFailed(let detail):
      return .downloadFailed(detail)
    case .downloadTimeout(let detail):
      return .downloadTimeout(detail)
    }
  }

  // Resolves absolutePath, or directory + fileName (UUID default).
  private static func resolveOutputPath(output: OutputConfig?) throws -> String {
    if let absolutePath = output?.absolutePath, !absolutePath.isEmpty {
      let parentDirectory = (absolutePath as NSString).deletingLastPathComponent
      var isDirectory: ObjCBool = false
      guard FileManager.default.fileExists(atPath: parentDirectory, isDirectory: &isDirectory), isDirectory.boolValue else {
        throw PdfError.writeFailed("Parent directory does not exist: \(parentDirectory)")
      }
      guard FileManager.default.isWritableFile(atPath: parentDirectory) else {
        throw PdfError.permissionDenied("Parent directory is not writable: \(parentDirectory)")
      }
      return absolutePath
    }

    let directory = output?.directory ?? .cache
    let directoryURL: URL
    switch directory {
    case .cache:
      guard let url = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
        throw PdfError.writeFailed("Could not resolve the caches directory")
      }
      directoryURL = url
    case .documents:
      guard let url = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
        throw PdfError.writeFailed("Could not resolve the documents directory")
      }
      directoryURL = url
    case .temp:
      directoryURL = URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
    }

    if !FileManager.default.fileExists(atPath: directoryURL.path) {
      do {
        try FileManager.default.createDirectory(at: directoryURL, withIntermediateDirectories: true)
      } catch {
        throw PdfError.writeFailed("Failed to create output directory \(directoryURL.path): \(error.localizedDescription)")
      }
    }

    let fileName: String
    if let providedName = output?.fileName, !providedName.isEmpty {
      fileName = providedName
    } else {
      fileName = "\(UUID().uuidString).pdf"
    }

    // .atomic write overwrites an existing file at this path.
    return directoryURL.appendingPathComponent(fileName).path
  }
}

File: android/src/main/java/com/margelo/nitro/nitroimage2pdf/HybridImageToPdf.kt
package com.margelo.nitro.nitroimage2pdf

import android.content.Context
import android.util.Base64
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitroimage2pdf.CreatePdfOptions
import com.margelo.nitro.nitroimage2pdf.CreatePdfResult
import com.margelo.nitro.nitroimage2pdf.HybridImageToPdfSpec
import com.margelo.nitro.nitroimage2pdf.OutputConfig
import com.margelo.nitro.nitroimage2pdf.OutputDirectory
import com.margelo.nitro.nitroimage2pdf.OutputFormat
import com.margelo.nitro.nitroimage2pdf.PdfPageInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.io.File
import java.util.UUID

// Bridges HybridImageToPdfSpec to ImageLoader/PdfBuilder; handles output writing.
class HybridImageToPdf : HybridImageToPdfSpec() {
  private val context: Context
    get() = NitroModules.applicationContext
      ?: throw PdfException.WriteFailed("No application Context available from NitroModules")

  companion object {
    // IO dispatcher avoids starving Promise.async's default pool.
    private val ioScope = CoroutineScope(Dispatchers.IO)
  }

  override fun createPdf(options: CreatePdfOptions): Promise<CreatePdfResult> = Promise.async(ioScope) {
    if (options.images.isEmpty()) {
      throw PdfException.InvalidImage("images must be a non-empty array")
    }

    val appContext = context

    // Load every source image, downsampling per processing.maxWidth/maxHeight.
    val maxWidth = options.processing?.maxWidth?.toInt()
    val maxHeight = options.processing?.maxHeight?.toInt()
    val loadedImages = options.images.map { input ->
      ImageLoader.load(appContext, input, maxWidth, maxHeight)
    }

    // Build the PDF (page sizing, fit, margin, metadata, password).
    val (pdfBytes, pageSizes) = PdfBuilder.build(
      context = appContext,
      images = loadedImages,
      page = options.page,
      metadata = options.metadata,
      quality = options.processing?.quality,
      userPassword = options.security?.password,
      ownerPassword = options.security?.ownerPassword
    )

    // Write output per output.format/directory/absolutePath.
    val outputFormat = options.output?.format ?: OutputFormat.FILE
    val writeFile = outputFormat == OutputFormat.FILE || outputFormat == OutputFormat.BOTH
    val writeBase64 = outputFormat == OutputFormat.BASE64 || outputFormat == OutputFormat.BOTH

    var filePath: String? = null
    var fileSize: Double? = null

    if (writeFile) {
      val path = resolveOutputPath(appContext, options.output)
      try {
        File(path).writeBytes(pdfBytes)
      } catch (e: Exception) {
        throw PdfException.WriteFailed("Failed to write PDF to $path: ${e.message}")
      }
      filePath = path
      fileSize = pdfBytes.size.toDouble()
    }

    val base64: String? = if (writeBase64) Base64.encodeToString(pdfBytes, Base64.NO_WRAP) else null

    val pages = pageSizes.map { PdfPageInfo(width = it.width.toDouble(), height = it.height.toDouble()) }
    CreatePdfResult(
      filePath = filePath,
      base64 = base64,
      numberOfPages = loadedImages.size.toDouble(),
      fileSize = fileSize,
      pages = pages.toTypedArray()
    )
  }

  // Resolves absolutePath, or directory + fileName (UUID default).
  private fun resolveOutputPath(context: Context, output: OutputConfig?): String {
    val absolutePath = output?.absolutePath
    if (!absolutePath.isNullOrEmpty()) {
      val parent = File(absolutePath).parentFile
      if (parent == null || !parent.exists() || !parent.isDirectory) {
        throw PdfException.WriteFailed("Parent directory does not exist: ${parent?.path}")
      }
      if (!parent.canWrite()) {
        throw PdfException.PermissionDenied("Parent directory is not writable: ${parent.path}")
      }
      return absolutePath
    }

    val directory = output?.directory ?: OutputDirectory.CACHE
    val directoryFile = when (directory) {
      OutputDirectory.CACHE -> context.cacheDir
      OutputDirectory.DOCUMENTS -> context.filesDir
      OutputDirectory.TEMP -> context.cacheDir // same as cache on Android, unlike iOS
    } ?: throw PdfException.WriteFailed("Could not resolve output directory for $directory")

    if (!directoryFile.exists() && !directoryFile.mkdirs()) {
      throw PdfException.WriteFailed("Failed to create output directory ${directoryFile.path}")
    }

    val fileName = output?.fileName?.takeIf { it.isNotEmpty() } ?: "${UUID.randomUUID()}.pdf"

    // Overwrite-on-collision: `File.writeBytes` truncates + overwrites existing files.
    return File(directoryFile, fileName).path
  }
}
